### DISCUSSION TITLE
Questions about uri generations and scope

### DISCUSSION THREAD

**Discussion by mefenlon** [statamic-rad-pack/runway#439] [Help]
### Description

I am currently using runway as a read only interface for my resources. The resources I am using are a subest of all my laravel models. I was intially using global scopes to define these but have since used the `scopeRunwayListing`. This has led me to a few questions.

1. As I create new model instances in my laravel app outside of statamatic/runway, will I need to regurally run `runway:rebuild-uris` to capture the new items?
2. Is there a way to generate URIs on the fly without using a database table?
3. Since removing the global scopes, when I run `runway:rebuild-uris` I think it is generating a uri for all models in the database. Would it be helpful if I submitted a PR to limit this using the `scopeRunwayListing` like it does in control panel?

```
// src/Console/Commands/RebuildUriCache.php
                $this->withProgressBar($resource->model()->all(), function ($model) use ($resource) {
                    $uri = (new Parser())
                        ->parse($resource->route(), $model->toAugmentedArray())
                        ->__toString();

                    $uri = Str::start($uri, '/');

                    $model->runwayUri()->create([
                        'uri' => $uri,
                    ]);
                });
```
```
// src/Fieldtypes/BaseFieldtype.php
    public function getIndexItems($request)
    {
        $resource = Runway::findResource($this->config('resource'));

        $query = $resource->model()->orderBy($resource->orderBy(), $resource->orderByDirection());

        $query->when($query->hasNamedScope('runwayListing'), fn ($query) => $query->runwayListing());
        $query->when($request->search, fn ($query) => $query->runwaySearch($request->search));

        $items = $request->boolean('paginate', true)
            ? $query->paginate()
            : $query->get();
```

### Environment

Environment
Application Name: Ripm_Jazz
Laravel Version: 10.45.0
PHP Version: 8.2.15
Composer Version: 2.6.6
Environment: local
Debug Mode: ENABLED
URL: ripmjazz.test
Maintenance Mode: OFF

Cache
Config: NOT CACHED
Events: NOT CACHED
Routes: NOT CACHED
Views: CACHED

Drivers
Broadcasting: log
Cache: redis
Database: mysql
Logs: stack / single
Mail: smtp
Queue: sync
Session: file

Statamic
Addons: 2
Antlers: runtime
Sites: 1
Stache Watcher: Enabled
Static Caching: Disabled
Version: 4.49.0 PRO

Statamic Addons
jacksleight/statamic-bard-texstyle: 3.1.6
statamic-rad-pack/runway: 6.2.1

**Comment by duncanmcclean** [comment]
> As I create new model instances in my laravel app outside of statamatic/runway, will I need to regurally run runway:rebuild-uris to capture the new items?

You shouldn't need to. The `RunwayRoutes` trait has an event listener on it to regenerate the URLs when models are created or updated.

https://github.com/statamic-rad-pack/runway/blob/6.x/src/Routing/Traits/RunwayRoutes.php#L76-L100

> Is there a way to generate URIs on the fly without using a database table?

There's not - since you can define Runway routes with Antlers, Runway needs to save them in the database in order for them to be queried later. 

If you don't want them saved in the database, you could disable Runway's routing functionality and define your own routes in `routes/web.php` and do things yourself.

> Since removing the global scopes, when I run runway:rebuild-uris I think it is generating a uri for all models in the database. Would it be helpful if I submitted a PR to limit this using the scopeRunwayListing like it does in control panel?

The `scopeRunwayListing` query scope is only intended to filter down the models visible in the Control Panel listing table. 

For example: it could be used to only return models assigned to the logged in user or override the ordering of a model. Those aren't things you'd necessarily want to take affect elsewhere, like in the `{{ runway }}` tag.

However, if it'd be useful for you, I would accept a PR that adds another query scope just for the `runway:rebuild-uris` command, maybe called `scopeRunwayRoutes` or something similar?

**Reply by duncanmcclean** [reply]
For those who stumble across this discussion in the future, you can [now define a `runwayRoutes` scope](https://runway.duncanmcclean.com/frontend-routing#content-building-the-uri-cache) on your model to filter which models are included in the "URI cache".
